tempesta_fw/hpack.c

```diff
@@ -0,0 +1,3439 @@
+			tfw_hpack_free_entry(cp);
+			early++;
+			cp++;
+			count--;
+			if (unlikely(early == length)) {
+				early = 0;
+				cp = entries;
+			}
+		} while (size > new_size);
+
+		tbl->n = count;
+		tbl->size = size;
+	}
+
+	tbl->window = new_size;
+
+	return 0;
+}
+
+static inline void
+tfw_huffman_init(TfwHPack *__restrict hp)
+{
+	hp->curr = -HT_NBITS;
+}
+
+int
+tfw_hpack_init(TfwHPack *__restrict hp, unsigned int htbl_sz)
+{
+	TfwHPackETbl *et = &hp->enc_tbl;
+	TfwHPackDTbl *dt = &hp->dec_tbl;
+
+	BUILD_BUG_ON(sizeof(TfwHPackNode) > 32
+		     || HPACK_ENC_TABLE_MAX_SIZE > SHRT_MAX);
+
+	tfw_huffman_init(hp);
+
+	dt->window = hp->max_window = htbl_sz;
+	if (!(dt->pool = __tfw_pool_new(0)))
+		return -ENOMEM;
+
+	et->window = htbl_sz;
+	spin_lock_init(&et->lock);
+	et->rb_size = HPACK_ENC_TABLE_MAX_SIZE;
+	if (!(et->pool = __tfw_pool_new(HPACK_ENC_TABLE_MAX_SIZE))) {
+		tfw_pool_destroy(dt->pool);
+		return -ENOMEM;
+	}
+	et->rbuf = tfw_pool_alloc(et->pool, HPACK_ENC_TABLE_MAX_SIZE);
+	BUG_ON(!et->rbuf);
+
+	return 0;
+}
+
+void
+tfw_hpack_clean(TfwHPack *__restrict hp)
+{
+	tfw_pool_destroy(hp->enc_tbl.pool);
+	tfw_hpack_set_length(hp, 0);
+	tfw_pool_destroy(hp->dec_tbl.pool);
+	WARN_ON_ONCE(act_hp_str_n);
+}
+
+/*
+ * HPACK reinitialization procedure: resetting the non-permanent part
+ * of HPACK context and HTTP/2 message iterator before next HPACK
+ * processing stage. Note, in result of reinitialization, the @state
+ * field of HPACK context will be set to HPACK_STATE_READY (since its
+ * value is zero).
+ */
+static inline void
+tfw_hpack_reinit(TfwHPack *__restrict hp, TfwMsgParseIter *__restrict it)
+{
+	bzero_fast(hp->__off,
+		   sizeof(*hp) - offsetof(TfwHPack, __off));
+	bzero_fast(it->__off,
+		   sizeof(*it) - offsetof(TfwHPack, __off));
+	TFW_STR_INIT(it->hdr);
+}
+
+static int
+tfw_hpack_hdr_process(TfwHttpReq *req)
+{
+	int r = T_OK;
+	TfwMsgParseIter *it = &req->pit;
+	const TfwStr *chunk, *end, *hdr = it->hdr;
+
+	BUG_ON(TFW_STR_DUP(hdr));
+	if (TFW_STR_PLAIN(hdr)) {
+		chunk = hdr;
+		end = hdr + 1;
+		WARN_ON_ONCE(chunk != it->next);
+	} else {
+		chunk = it->next;
+		end = hdr->chunks + hdr->nchunks;
+		WARN_ON_ONCE(hdr == it->next);
+	}
+
+	while (chunk < end) {
+		r = tfw_h2_parse_hdr(chunk->data, chunk->len, req);
+		if (unlikely(r < T_POSTPONE))
+			return r;
+		++chunk;
+	}
+
+	return T_OK;
+}
+
+static bool
+tfw_hpack_huffman_parse(TfwHPack *__restrict hp, const unsigned char **src,
+			unsigned long *n, TfwMsgParseIter *__restrict it)
+{
+	/*
+	 * TODO #309: implementation of Huffman parser; inside this procedure
+	 * we also should write to prepared header storage the plain constant
+	 * matched header names, ':' symbol after the name, and set header's ID
+	 * tag.
+	 */
+	return false;
+}
+
+/*
+ * HPACK decoder FSM for HTTP/2 message processing.
+ */
+int
+tfw_hpack_decode(TfwHPack *__restrict hp, const unsigned char *src,
+		 unsigned long n, TfwHttpReq *__restrict req,
+		 unsigned int *__restrict parsed)
```

Comment:
Yes - agree.

Reply:
I have written a lot of code of encoding logic in last PR, thus it seems that it will be more effective to make this subtask in context of last PR - to avoid hard merging and numerous conflicts' resolution.